text/3334-niche.md

```diff
@@ -0,0 +1,348 @@
+- Start Date: 2022-10-16
+- RFC PR: [rust-lang/rfcs#3334](https://github.com/rust-lang/rfcs/pull/3334)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Provide a stable attribute to define "niche" values of a type. The type cannot
+store these values, allowing the compiler to use them to optimize the
+representation of containing types such as `Option<Type>`.
+
+# Motivation
+[motivation]: #motivation
+
+Rust makes extensive use of types like `Option`, and many programs benefit from
+the efficient storage of such types. Many programs also interface with others
+via FFI, via interfaces that provide data and a sentinel value (such as for
+errors or missing data) within the same bits.
+
+The Rust compiler already provides support for this via "niche" optimizations,
+and various types providing guarantees of such optimizations, including
+references, `bool`, `char`, and the `NonZero` family of types. However, Rust
+does not provide any stable means of defining new types with niches, reserving
+this mechanism for the standard library. This puts pressure on the standard
+library to provide additional families of types with niches, while preventing
+the broader crate ecosystem from experimenting with such types.
+
+Past efforts to define a stable niche mechanism stalled out due to scope creep:
+alignment niches, null-page niches, multiple niches, structures with multiple
+fields, and many other valid but challenging ideas (documented in the "Future
+possibilities" section). This RFC defines a *simple* mechanism for defining one
+common type of niche, while leaving room for future extension.
+
+Defining a niche mechanism allows libraries to build arbitrary types containing
+niches, and simplifies handling of space-efficient data structures in Rust
+without manual bit-twiddling.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a structure containing exactly one field of a non-zero-sized type
+(non-ZST), you can attach a `niche` attribute on it to declare a specific value
+or range of values for that field as invalid. This promises the compiler that
+you will never store those values in that field, which allows the compiler to
+use those in-memory representations for different purposes, such as the
+representation of `None` in a containing `Option`.
+
+```rust
+use std::mem::size_of;
+
+#[niche(value = 42)]
+struct MeaninglessNumber(u64);
+
+assert_eq!(size_of::<MeaninglessNumber>(), 8);
+assert_eq!(size_of::<Option<MeaninglessNumber>>(), 8);
+
+#[niche(range = 2..)]
+struct Bit(u8);
+
+assert_eq!(size_of::<Bit>(), 1);
+assert_eq!(size_of::<Option<Option<Option<Bit>>>>(), 1);
+```
+
+Constructing a structure with a niche value, or writing to the non-ZST field of
+such a structure, or obtaining a mutable reference to such a field, requires
+`unsafe` code. Causing a type with a niche to contain an invalid value (whether
```

Comment:
I'm highly skeptical of this, because it feels like it's repeating the same mistakes that rust ended up having to spend multiple years fixing for `repr(packed)` (culminating finally in https://github.com/rust-lang/rust/pull/102513).

Namely, you now have something supposedly "typed" as `u32`, but it's not really a `u32` at all, because it has extra constraints.  Sure, you can say that you need to be an `unsafe` block to get a `&mut u32` to it, but that doesn't fix the "this is the wrong place to put the proof obligation" problem.  Obviously this isn't *as* bad as `mem::uninitialized`, but it has the same structural problems: asking someone to say "I promise that nothing after here will ever misuse this" is awkward, compared to only asking "when you construct a value of the type, you have to prove that it's in-range".

This seems awkward for MIRI to check, too.  Because if you have a `&mut u32`, how is it going to know that it's illegal to write a `42_u32` through it?  Where does the UB happen?  Since we don't have typed memory, it seems like it would be unable to diagnose the write as UB, and would have to wait until something later read it, which seems suboptimal -- especially for types that are `Copy` and thus might not ever be read!

For example, is this UB?

```rust
#[niche(value = 42)]
#[derive(Copy, Clone)]
struct MeaninglessNumber(u64);

unsafe {
    let mut x = Some(MeaninglessNumber(10));
    let r: &mut u64 = &mut x.as_mut().unwrap_unchecked().0;
    *r = 42;
}
```

It feels to me like it *ought* to be UB -- conceptually it's setting the value to its niche, which is definitely not okay -- but operationally I don't see *where* we can diagnose the problem.

I'm very much in favour of letting people control niches, but I think u32-with-a-particular-niche and normal u32 need to be different types.

Reply:
> It feels to me like it ought to be UB

There's a lot of discussion of this in https://github.com/rust-lang/unsafe-code-guidelines/issues/84. It's been a while but I think the consensus is that it's very hard to enforce, we don't gain much from making it UB (we can have validity enforced on typed copy and load), and there are valid use cases for it.

Reply:
@scottmcm thank you for bringing this up, it's important that this is discussed somewhere. However, there is a nice way to represent this in the opsem. The interesting minirust operation here is a "load" - the operation that turns a list of bytes into a value at the given type. For tuples (and structs), that is currently defined like this:

```rust
fn decode(Type::Tuple { fields, size }: Self, bytes: List<AbstractByte>) -> Option<Value> {
    if bytes.len() != size { throw!(); }
    Value::Tuple(
        fields.iter().map(|(offset, ty)| {
            ty.decode(bytes[offset..][..ty.size()])
        }).collect()?,
    )
}
```

Essentially, this is saying that a value consists of all the values of the fields, and that the load is DB if the loads of all the fields are DB. Under this proposal, I believe we would adjust it to be the following:

```rust
fn decode(Type::Tuple { fields, size }: Self, bytes: List<AbstractByte>) -> Option<Value> {
    if bytes.len() != size { throw!(); }
    Value::Tuple(
        fields.iter().map(|(offset, ty, niche)| {
            let field_val = ty.decode(bytes[offset..][..ty.size()])?;
            if Some(niche) = niche {
                if !niche.contains(field_val.unwrap_primitive_int()) {
                    throw_ub!();
                }
            }
        }).collect()?,
    )
}
```

This additionally checks that if the field has a niche annotation, then the value of that field is inside the niche range (this should also help explain why I feel so strongly about the fields getting niche annotations, not the type).

Of course this only matters when there is a memory operation at type `Foo`. Your example above does not contain any such memory operation, and so it has no UB. Even something like this would be permitted:

```rust
struct Foo(
    #[niche(range = 4..8)]
    u8,
);

let mut x: Foo = Foo(5);
x.0 = 0;
dbg!(x.0);
```

The memory operations are at type `u8`, and so there is no UB. If it were instead to say `dbg!(x)` then there would be a load at type `Foo` and we would have UB.

As far as I know, these semantics are completely reasonable from an opsem perspective. It of course remains to be answered if they are the semantics we want, but I can't think of any concrete reason we should disfavor them.

Reply:
Thanks, @thomcc, that's a good read.

I feel like there's a difference between references and pointers, here, though.

For example, if I had to write that as
```rust
unsafe {
    let mut x = Some(MeaninglessNumber(10));
    let p = ptr::addr_of_mut!(x.as_mut().unwrap_unchecked().0);
    p.cast::<u64>().write(42);
}
```
then I agree that it's logical that there's no UB, same as if I'd done just `x.as_mut().unwrap_unchecked() as *mut _`.

Perhaps it's illogical, but I expect more restrictions from references, and expect to have to fall down to pointers to do squicky things like changing the variant of an option by writing through reference to the field.

Reply:
I don't think this is really about this RFC anymore, so [moved to Zulip](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/Changing.20Variants.20Through.20References/near/305723808).

Reply:
I realized I forgot to say part of this: if the type inside `MeaninglessNumber` was different from `u64`, then it *would* be difference.  If you have a `&mut Not42`, then you can't write 42 through that because it's a typed write.  If you turn it into a pointer, though, and `.cast::<u64>()`, then sure, you can do weird things, I agree.

I just don't think we should be giving out a `&mut u64`, as this RFC seems like it does, instead of a `&mut Not42` .

Reply:
I agree. Niches need to be part of the type system, we can't keep bolting things on the side like `pached`.

Reply:
> I just don't think we should be giving out a &mut u64, as this RFC seems like it does, instead of a &mut Not42 .

Just rejecting giving out these references is also an option. But we if allow giving them out, we have to make it `unsafe` -- I hope that much is uncontroversial. :)  (And that's what the current rustc-private niche attribute does.)

FWIW this is not quite as bad as `packed` since creating an `&mut u64` reference that points to 42 is obviously perfectly fine in and of itself. It's only *writes* to that reference that are subtle, and that's where we hit https://github.com/rust-lang/unsafe-code-guidelines/issues/84.

Reply:
To the extent this is related to the RFC, a few points:

1. I do think we should support giving out a `&mut FieldType` *from unsafe code*, precisely because this allows calling other functions that mutate a field that were designed to operate on the field type. Otherwise, if you want to mutate the field, you have to extract it, operate on it, and write the result back to it. This seems to me like a reasonable operation for unsafe code to have. This isn't at all like `packed`, as @RalfJung noted. With packed, you can end up with an unaligned reference; with niches you cannot.

2. Even if we want to provide types that use const generics to represent value ranges, we need some means of *creating* those types. I would like to give the ecosystem the tools we're using, not just things we've built with those tools. (See the motivation of the RFC.)

Reply:
I've added a more detailed discussion of mutable references. I'm going to mark this "resolved" for now, but feel free to un-resolve if there's anything missing.

Reply:
I'm unresolving because to me this isn't so much about about the soundness rules -- I agree it's *sound* to give out references here, in the same way that it's *sound* to give out `&mut u32`s to `f32`s -- but because the field type is still basically a lie.

Completely by happenstance, I just stumbled on an interesting example of how the "`u32` but not really a `u32`" does weird things: https://github.com/rust-lang/rust/pull/103556#issuecomment-1291538814.

For the 2 years that git will conveniently trace history (https://github.com/rust-lang/rust/blame/e2267046859c9ceb932abc983561d53a117089f6/library/core/src/num/nonzero.rs#L47) and probably forever, we've been deriving `PartialEq` for the `NonZero` types.  But what does `derive` do for a `struct`?  Takes references to the fields and calls `PartialEq` using those references.

And voila, the type lie immediately impacts things.  Rather than telling LLVM we're comparing non-zero types by emitting the `!range` metadata that would let it take advantage of the `rustc_layout_scalar_valid_range_start` -- as happens if you write the obvious implementation yourself <https://rust.godbolt.org/z/Ye5xr8P8x> -- it thinks it's loading ordinary integers and the resulting code has no range metadata.  That missing metadata then, presumably, makes it hard for LLVM to optimize `==` on `Option<NonZeroNN>`, resulting in the linked PR to use specialization to try to fix it.

So then this very much *is* liked `packed` in the sense that if you want your derive macro to handle these types well then you'll have to look for it specifically in the macro and handle those cases differently (https://github.com/rust-lang/rust/pull/98446/files#diff-d1ddb1b94e4a7a467f9a396742ceaa3a94e7cdc96055c36a18b8116614682df1R965-R968).

Is that *really* worth it to be able to write
```rust
#[niche = 4]
struct MyType(u32);
```
instead of a normal newtype?
```rust
#[repr(transparent)]
struct MyType(NicheU32<4>);
```
Personally, I don't think so.

See also, for example, `derive(Deserialize)`.  For the `#[niche]` that's something that either won't work -- because it needs `unsafe` to construct -- or needs special handling in the derive macro.  Whereas for the newtype it just needs a library implementation for the `NicheU32` type, and I'll take normal rust library code over a bespoke path in the proc macro any day.

Reply:
> I agree it's sound to give out references here, in the same way that it's sound to give out &mut u32s to f32s

FWIW it is definitely not *sound* to give out `&mut u32` to a `NonZeroU32`, in the sense [defined here](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#soundness-of-code--of-a-library).

But it is possible to write UB-free code that takes an `&mut u32` to a `NonZeroU32`.